### PR TITLE
Make moving Shapes, swapping Textures and swapping VertexBuffers noexcept

### PR DESCRIPTION
**Pull request by binary1248**
As reported at https://sonarcloud.io/project/issues?resolved=false&severities=BLOCKER&id=SFML_SFML certain move operations and swaps weren't noexcept. The C++ core guidelines recommend these types of operations should never fail because they really shouldn't in properly designed code.

This changeset fixes these issues.

Additional points:
- In order to make `sf::Texture::swap()` noexcept, `getUniqueId()` also has to be noexcept, the new implementation uses an atomic counter
- In order to use certain atomic features, GCC requires explicitly linking libatomic when building on certain platforms e.g. armhf
- A `std::swap` test for `sf::Texture` was not added because it would cause too many tests to fail on Windows due to OpenGL issues (hopefully will be fixed at some point)

### CODE REVIEW

File: cmake/Macros.cmake [hunk 1]
```diff
@@ -35,6 +35,17 @@ function(sfml_set_stdlib target)
...
+    # on some platforms (e.g. ARM), GCC requires linking libatomic to use <atomic> features
+    if(SFML_COMPILER_GCC)
+        include(CheckCXXSourceCompiles)
+        check_cxx_source_compiles("#include <atomic>
+            #include <cstdint>
+            int main(){std::atomic<std::uint64_t> x(1); return x.fetch_add(1);}" ATOMIC_TEST)
+        if(NOT ATOMIC_TEST)
+            target_link_libraries(${target} PUBLIC "-latomic")
+        endif()
+    endif()
```
Comment (ChrisThrasher) [inline cmake/Macros.cmake:48]: I frequently use GCC on ARM and have never seen this before. What versions of GCC on what platforms are giving you issue?

Comment (binary1248) [reply cmake/Macros.cmake:48]: [Our own armhf worker](https://ci.sfml-dev.org/#/builders/29/builds/1864/steps/5/logs/stdio).

File: cmake/Macros.cmake [hunk 2]
```diff
@@ -35,6 +35,17 @@ function(sfml_set_stdlib target)
             message(FATAL_ERROR "Clang is the only supported compiler on macOS")
         endif()
     endif()
+
+    # on some platforms (e.g. ARM), GCC requires linking libatomic to use <atomic> features
```
Comment (ChrisThrasher) [inline cmake/Macros.cmake:39]: Can we clarify that it's specific armhf/RasPi that is the main motivator here?

File: cmake/Macros.cmake [hunk 3]
```diff
@@ -35,6 +35,17 @@ function(sfml_set_stdlib target)
...
+    if(SFML_COMPILER_GCC)
```
Comment (ChrisThrasher) [inline cmake/Macros.cmake:40]: Does this same problem apply to those using Clang on Raspberry Pis?

File: cmake/Macros.cmake [hunk 4]
```diff
@@ -35,6 +35,17 @@ function(sfml_set_stdlib target)
...
+        include(CheckCXXSourceCompiles)
+        check_cxx_source_compiles("#include <atomic>
+            #include <cstdint>
+            int main(){std::atomic<std::uint64_t> x(1); return x.fetch_add(1);}" ATOMIC_TEST)
```
Comment (ChrisThrasher) [inline cmake/Macros.cmake:44]: Minor knit but if you replaced `std::uint64_t` with `int` you could drop `<cstdint>`. Presumably it's `fetch_add` that causes issues, not that we used a 64 bit unsigned int.

File: include/SFML/Graphics/VertexBuffer.hpp [hunk 5]
```diff
@@ -338,6 +338,15 @@ class SFML_GRAPHICS_API VertexBuffer : public Drawable, private GlResource
...
 };
 
+////////////////////////////////////////////////////////////
+/// \brief Swap the contents of one vertex buffer with those of another
+///
+/// \param left First instance to swap
+/// \param right Second instance to swap
+///
+////////////////////////////////////////////////////////////
+void swap(VertexBuffer& left, VertexBuffer& right) noexcept;
```
Comment (ChrisThrasher) [inline include/SFML/Graphics/VertexBuffer.hpp:348]: Why did you add this free function? Does it mean we should make `sf::VertexBuffer::swap` a private method that is merely an implementation detail of this overload?

File: test/Graphics/RectangleShape.test.cpp [hunk 6]
```diff
@@ -7,8 +7,8 @@
 
 static_assert(std::is_copy_constructible_v<sf::RectangleShape>);
 static_assert(std::is_copy_assignable_v<sf::RectangleShape>);
-static_assert(std::is_move_constructible_v<sf::RectangleShape>);
-static_assert(std::is_move_assignable_v<sf::RectangleShape>);
+static_assert(std::is_nothrow_move_constructible_v<sf::RectangleShape>);
+static_assert(std::is_nothrow_move_assignable_v<sf::RectangleShape>);
```
Comment (ChrisThrasher) [inline test/Graphics/RectangleShape.test.cpp:11]: 🥳

File: cmake/Macros.cmake [hunk 7]
```diff
@@ -35,6 +35,17 @@ function(sfml_set_stdlib target)
...
+        if(NOT ATOMIC_TEST)
+            target_link_libraries(${target} PUBLIC "-latomic")
```
Comment (ChrisThrasher) [inline cmake/Macros.cmake:46]: Because we're only using this in Texture.cpp, can we link this as a `PRIVATE` dependency? I don't see why this needs to become a transitive dependency. Also, can't this merely be a link library for the graphics module only instead of a link library for every single module?

**Review by ChrisThrasher — COMMENTED** [review]
I like this. Just a few comments about implementation tweaks then I'm good to approve.

File: cmake/Macros.cmake (hunk 3 above)
Comment (binary1248) [reply cmake/Macros.cmake:40]: I've only seen people reporting about this using GCC since it was an active decision of theirs to split the atomic implementation out of the standard runtime library that is automatically linked in. The fact that this doesn't happen when building with GCC for aarch64 (RPi 64-bit) makes me think they regret this decision and have just followed what all other sane compilers do.

File: cmake/Macros.cmake (hunk 4 above)
Comment (binary1248) [reply cmake/Macros.cmake:44]: Depending on whether the operation is translated into a compiler intrinsic or a function implementation, linking to the external library might not be required. Since this is what we use in the code, I figured it is safer to test for this exact code even though your version would be shorter. I can build in the shorter version and see what happens.

File: include/SFML/Graphics/VertexBuffer.hpp (hunk 5 above)
Comment (binary1248) [reply include/SFML/Graphics/VertexBuffer.hpp:348]: `std::is_nothrow_swappable_v` requires a non-member swap being available. There are people on both sides of the fence on whether to have only 1 or both implementations, but I personally don't see why we can't just keep both especially since removing the member function at this point would just be unnecessarily causing API breakage without much benefit.

File: cmake/Macros.cmake (hunk 7 above)
Comment (binary1248) [reply cmake/Macros.cmake:46]: We have to be careful here. Currently `<atomic>` is only used in Texture.cpp. Because it is likely that we might use `<atomic>` in more places as we modernize the code and replace stuff with `atomic<T>`s, I wouldn't want to have to copy and paste the same block of code into multiple places. This is a very unusual library because it is needed to support standard library features, as such I wouldn't put them in the same basket as the other libraries. Depending on whether we end up using `<atomic>` in a header somewhere, the library might also have to be made public. I can change it to private for now if you want, but there is a possibility we will have to make it public again some time in the future.

Comment (ChrisThrasher) [reply cmake/Macros.cmake:46]: > I wouldn't want to have to copy and paste the same block of code into multiple places.

You'll never need to do that. As soon as we need this in two module we'll put it in some higher common location.

> there is a possibility we will have to make it public again some time in the future.

This is perfectly fine. We'll make that change when the time comes.

**Review by ChrisThrasher — APPROVED** [review]

File: include/SFML/Graphics/Shape.hpp [hunk 8]
```diff
@@ -52,6 +52,30 @@ class SFML_GRAPHICS_API Shape : public Drawable, public Transformable
     ////////////////////////////////////////////////////////////
     ~Shape() override;
```
Comment (ChrisThrasher) [inline include/SFML/Graphics/Shape.hpp:53]: https://godbolt.org/z/cYs4exPs3

@kimci86 Brought up how we can apparently remove `~Shape` which will implicit define all these other special member functions. Because `sf::Shape` inherits from a class that has a virtual distractor, it too inherits a virtual destructor. The above Godbolt link shows that this won't introduce memory leaks. We can use `std::has_virtual_destructor_v<T>` to assert this fact and ensure regressions don't slip in.

**Comment by binary1248** [comment]
Fixed.

**Review by ChrisThrasher — APPROVED** [review]
😊

File: src/SFML/Graphics/VertexBuffer.cpp [hunk 9]
```diff
@@ -333,22 +330,15 @@ VertexBuffer::Usage VertexBuffer::getUsage() const
...
-
         TransientContextLock contextLock;
 
         // Make sure that extensions are initialized
         sf::priv::ensureExtensionsInit();
 
-        available = GLEXT_vertex_buffer_object;
-    }
+        return GLEXT_vertex_buffer_object;
+    }();
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/VertexBuffer.cpp:341]: My C++-fu isn't the best here, does the static guarantee some atomicy, or why is the mutex no longer needed?

Comment (binary1248) [reply src/SFML/Graphics/VertexBuffer.cpp:341]: https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

**Review by eXpl0it3r — APPROVED** [review]

**Comment by ChrisThrasher** [comment]
Thanks! Now let's see if SonarCloud is happier with these changes.
